File: Threads/app/src/main/java/com/example/threads/MainActivity.java
package com.example.threads;

import androidx.annotation.LongDef;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {

    private Button botaoIniciar;
    private Button botaaoParar;
    private int numero;
    private Handler handler = new Handler();
    private Boolean pararExecucao = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        botaoIniciar = findViewById(R.id.buttonIniciar);
        botaaoParar = findViewById(R.id.buttonParar);

    }

    public void iniciarThread(View view) throws InterruptedException {

        /*MyThread thread = new MyThread();
        thread.start();
         */
        pararExecucao = false;
        MyRunnable runnable = new MyRunnable();
        new Thread( runnable ).start();

        /*
        Sobrecarrega a UI Thread
        for (int i=0; i<=15; i++){
            Log.d("thread", "contador: " + i );
            try {
                 Thread.sleep(1000);
            } catch (InterruptedException e){
                e.printStackTrace();
            }
        } */
    }

    public void pararThread(View view){

        pararExecucao = true;
    }

    class MyThread extends Thread {

        @Override
        public void run() {
            // cria nova Thread e não sobrecarrega a UI Thread
            for (int i=0; i<=15; i++){
                Log.d("thread", "contador: " + i );
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e){
                    e.printStackTrace();
                }
            }
            super.run();
        }
    }

    class MyRunnable implements Runnable {

        @Override
        public void run() {
           /* for (int i=0; i<=15; i++){
                numero = i;
                Log.d("thread", "contador: " + i );
                //botaaoIniciar.setText("Contador: " + i );

            */
                /*
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        botaaoIniciar.setText("Contador: " + numero );
                    }
                });
                 */

            for (int i = 0; i <= 15; i++) {
                if (pararExecucao){
                    return;
                }
                numero = i;
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e){
                    e.printStackTrace();
                }
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        botaoIniciar.setText("Contador: " + numero);
                    }
                });

            }
        }

    }
}

File: NavigationAndThreads/app/src/main/java/com/example/navigationandthreads/activities/MainActivity.kt
package com.example.navigationandthreads.activities



import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import com.example.navigationandthreads.Usuario
import com.example.navigationandthreads.databinding.ActivityMainBinding
import kotlinx.coroutines.*
import java.lang.Runnable

class MainActivity : AppCompatActivity() {

    private val binding by lazy {
        ActivityMainBinding.inflate(layoutInflater)
    }

    private var pararThread = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        binding.buttonAbrirTela.setOnClickListener{ abrirActivityComFragments() }
        binding.buttonIniciar.setOnClickListener{ iniciarCorroutine() }
        binding.buttonParar.setOnClickListener{ pararThread()}
    }

    private fun iniciarCorroutine(){
        CoroutineScope( Dispatchers.IO ).launch {
            /*val executou: String = getString(R.string.executou)
            binding.buttonIniciar.text = executou*/

           /* repeat(30){indice ->
                Log.i("info_thread", "Executando: $indice T:${Thread.currentThread().name}")
                withContext(Dispatchers.Main){
                    binding.buttonIniciar.text = "executando $indice"
                }
                delay(1000)

            }*/
           executarSuspensa()


        }
    }

    private suspend fun executarSuspensa(){
        val usuario = recuperarUsuarioLogado()
        Log.i("info", "Usuario: ${usuario.nome} ${Thread.currentThread().name}")
        val postagens = recuperarPostagemId( usuario.id )
        Log.i("info", "Postagens: ${postagens.size} ${Thread.currentThread().name}")
    }

    private suspend fun recuperarPostagemId(idUsuario: Int): List<String> {
        delay(2000)
        return listOf(
            "Estudando física",
            "Estudando Android"
        )
    }

    private suspend fun recuperarUsuarioLogado(): Usuario {
        delay(2000)
        return Usuario(1020, "Paulo Carpinetti")
        Log.i("info", "Recuperou Usuario!${Thread.currentThread().name}")
    }

    private fun pararThread(){
        pararThread = true
        iniciarBotao()
    }

    private fun iniciarThread() {
        //MinhaThread().start()
        Thread( MinhaRunnable() ).start()
    }

    private fun iniciarBotao(){
        binding.buttonIniciar.text = "Reiniciar execução"
        binding.buttonIniciar.isEnabled = true
    }

    private fun abrirActivityComFragments() {
        val intent = Intent(this, SegundaActivity::class.java)
        startActivity( intent )
        Log.i("info", "Thread:${Thread.currentThread().name}")

    }

    inner class MinhaRunnable : Runnable {
        override fun run() {

            if (pararThread) {
                pararThread = false
                return
            }

            repeat(30){indice ->
                Log.i("info_thread", "Executando: $indice T:${Thread.currentThread().name}")
                Thread.sleep(1000)
                runOnUiThread{
                    binding.buttonIniciar.text = "executando $indice"
                    binding.buttonIniciar.isEnabled = false
                    if ( indice == 29){
                        iniciarBotao()
                    }
                }
                Thread.sleep( 1000 )
            }
        }

    }
    inner class MinhaThread : Thread() {

        override fun run() {
            super.run()

            if (pararThread)return

            repeat(30){indice ->
                Log.i("info_thread", "Executando: $indice T:${currentThread().name}")
                sleep(1000)
                runOnUiThread{
                    binding.buttonIniciar.text = "executando $indice"
                    binding.buttonIniciar.isEnabled = false
                    if ( indice == 29){
                        iniciarBotao()
                    }
                }
                sleep( 1000 )
            }

        }
    }


}